src/java/org/apache/cassandra/db/commitlog/CommitLogReplayer.java

```diff
@@ -436,16 +455,23 @@ public void handleMutation(Mutation m, int size, int entryLocation, CommitLogDes
             sawCDCMutation = true;
 
         pendingMutationBytes += size;
+
+        boolean isSchemaMutation = SchemaConstants.isSchemaKeyspace(m.getKeyspaceName());
+
+        if (isSchemaMutation)
+            writeOrder.awaitNewBarrier();
```

Comment:
Would we want to put this one on a separate method, then on the dtest have a byteman counter and check that on the test?

Now there's just a delay which seems pretty nondeterministic. Counting how many schema changes we issue vs counter should be more deterministic imo.

Reply:
I don't understand how counting schema mutations would help us to verify that we are reproducing the problematic case, since we are not sending more mutations of this kind and the problem is due to the order in which mutations are replayed. Also, regular node starts include multiple schema mutations (~20) besides the ones that we introduce with the test, and it's not easy to identify which ones are the relevant ones with byteman.

I think that the induced 2 seconds delay is a really long time for replaying mutations, and it seems enough to guarantee that not-schema mutations are replayed before schema mutations. Indeed, the test passes [1000 iterations against the patched branch](https://app.circleci.com/pipelines/github/adelapena/cassandra/1246/workflows/9ab6f30b-8e87-486f-a5a7-2d1e74873ea5), while these other [1000 runs against unpatched trunk](https://app.circleci.com/pipelines/github/adelapena/cassandra/1248/workflows/03e75af8-985a-42ea-84f5-cf5f7480d6a7) fail every single time. 

However, running the test against trunk fails after the first replay, which doesn't even need the byteman rule. I we slightly modify the test to skip the first replay and go straight to the second replay with the byteman rule, [this way](https://github.com/adelapena/cassandra-dtest/commit/d54cc4878d674d3736bd71c88ae192f08aa9a622), the modified tests fails [1000 runs](https://app.circleci.com/pipelines/github/adelapena/cassandra/1250/workflows/75ee8eba-c7a7-4e7c-96bb-30e82d2f4520) against trunk, while it succeeds another [1000 runs](https://app.circleci.com/pipelines/github/adelapena/cassandra/1249/workflows/9182133c-aee0-4b58-bf1b-8082ff3a3cbf) against the patched branch. 

I thinks these runs show the stability of the test. We could try to add some locking with a couple of byteman rules to be able to somehow produce the desired ordering of replayed mutations, more or less reverting the action of the introduced `writeOrder`, but I think that would introduce unneeded complexity, making the rules more sensitive to future changes in `CommitLogReplayer`. wdyt?

Reply:
I'm pointing at in between each of these:

```
        session.execute("INSERT INTO ks.t (k, a, b) VALUES(2, 20, 100)")
        session.execute("ALTER TABLE ks.t ADD c int")
        session.execute("INSERT INTO ks.t (k, a, b, c) VALUES(3, 30, 300, 3000)")
        session.execute("ALTER TABLE ks.t DROP b")
        session.execute("INSERT INTO ks.t (k, a, c) VALUES(4, 40, 4000)")
```

Checking the counter increased would guarantee schemas mutations are _indeed_ detected and the 'wait' happens. Does it make more sense now? I'm not feeling strongly about it but it's sthg that crossed my mind.

Reply:
I see. Those queries wouldn't immediately increase the counter. That counter would be increased later, during the batchlog replay phase of the node restart that comes after those lines. The problem is that the batchlog replay of the default schema includes a bunch unrelated schema mutations, and we don't have an easy way to identify our mutations of interest. Besides, I don't think that Python dtests have the tooling to easily introduce that counter. 

What we can do instead is simply emit a log message when applying the rule, and then make the test look into the logs to verify that the injected delay is being applied, [this way](https://github.com/adelapena/cassandra-dtest/commit/635a5ef4686a5e2b1d761b648e530b7907bc099a). This would allow us to quickly detect in the future that we need to update the rule.

Reply:
Nice thinking about adding logging to the rule I'll try to remember that. This will check for any schema reload that might happen, not for the commit log replay specific ones correct? But I don't think there's a way to tell them apart even if we moved the barrier await to it's own method. So this is better than nothing imo.

Reply:
Yes, this will verify any schema reload, and starting the node would involve multiple schema reloads (and delays) because all the schema is replayed, not only our changes. We don't have an easy way to distinguish our own changes in the schema but we know that we are delaying all schema updates. 

As for knowing that our schema changes are being replayed, this is implicitly verified by the last query, which is based on the new schema. To be sure we can do the query with and without explicit column names, so we know that the table has the right columns after replaying the schema mutations, [this way](https://github.com/apache/cassandra-dtest/pull/155/commits/7532f8ab939d26daa2cdcc024e67d47f872037dd).